xformers/components/multi_head_dispatch.py

```diff
@@ -143,8 +143,12 @@ def forward(
         self._check(value, "value")
         self._check(key, "key")
 
-        B, S_Q, _ = query.size()  # Batch x Sequence x Embedding (latent)
-        _, S_K, _ = key.size()  # K, Q's sequence length could differ
+        B, S_Q, EMB = (
+            query.size(0),
+            query.size(1),
+            query.size(2),
+        )  # Batch x Sequence x Embedding (latent) x ...
+        S_K = key.size(1)  # K, Q's sequence length could differ
```

Comment:
we could just use "[expand](https://pytorch.org/docs/stable/generated/torch.Tensor.expand.html#torch.Tensor.expand)" here (not repeat, else memory hog..), in case one of the batch size is not matching. This would have to go with a unit test, to ensure coverage :)

Reply:
Hmm, I think this is a good idea. I guess the logic is not quite obvious to me since we could have any combination of QKV has batch_size=1, others have batch_size != 1 so we'd need to do some weird checks to make it work, right? Something like:

```python
if query.size(0) == 1 and not (key.size(0) == 1 and value.size(0) == 1):
    query = query.expand_as(...)
elif:
    ...
```

This seems pretty messy. Is there a simpler way to "expand as the largest in dim 0"?

Reply:
Hmm, also, the `expand` call will break if the sequence lengths differ between Q and K, right?